sqlglot/optimizer/optimizer.py

```diff
@@ -25,19 +41,18 @@ def optimize(expression, schema=None, db=None, catalog=None):
                 3. {catalog: {db: {table: {col: type}}}}
         db (str): specify the default database, as might be set by a `USE DATABASE db` statement
         catalog (str): specify the default catalog, as might be set by a `USE CATALOG c` statement
+        rules (typing.Sequence): sequence of optimizer rules to use
+        **kwargs: If a rule has a keyword argument with a same name in **kwargs, it will be passed in.
     Returns:
         sqlglot.Expression: optimized expression
     """
+    possible_kwargs = {"db": db, "catalog": catalog, "schema": schema, **kwargs}
     expression = expression.copy()
-    expression = qualify_tables(expression, db=db, catalog=catalog)
-    expression = isolate_table_selects(expression)
-    expression = qualify_columns(expression, schema)
-    expression = pushdown_projections(expression)
-    expression = normalize(expression)
-    expression = unnest_subqueries(expression)
-    expression = expand_multi_table_selects(expression)
-    expression = pushdown_predicates(expression)
-    expression = optimize_joins(expression)
-    expression = eliminate_subqueries(expression)
-    expression = quote_identities(expression)
+    for rule in rules:
+        rule_kwargs = {
+            param: possible_kwargs[param]
+            for param in rule.__code__.co_varnames
```

Comment:
what the heck is this

Reply:
I added a little comment.
Unresolve if you're not happy with this magic.